crates/astria-sequencer/src/ibc/recover_client.rs

```diff
@@ -0,0 +1,124 @@
+use astria_core::protocol::transaction::v1::action;
+use astria_eyre::{
+    anyhow_to_eyre,
+    eyre::{
+        ensure,
+        OptionExt as _,
+        Result,
+        WrapErr as _,
+    },
+};
+use cnidarium::StateWrite;
+use penumbra_ibc::component::{
+    ClientStateReadExt as _,
+    ClientStateWriteExt as _,
+    ClientStatus,
+    ConsensusStateWriteExt as _,
+};
+
+use crate::{
+    app::{
+        ActionHandler,
+        StateReadExt as _,
+    },
+    authority::StateReadExt as _,
+    transaction::StateReadExt as _,
+};
+
+#[async_trait::async_trait]
+impl ActionHandler for action::RecoverClient {
+    async fn check_stateless(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn check_and_execute<S: StateWrite>(&self, mut state: S) -> Result<()> {
+        let from = state
+            .get_transaction_context()
+            .expect("transaction source must be present in state when executing an action")
+            .address_bytes();
+        ensure!(
+            from == state
+                .get_sudo_address()
+                .await
+                .wrap_err("failed to get sudo address")?,
+            "only the sudo address can recover a client",
+        );
+
+        let timestamp = state
+            .get_block_timestamp()
+            .await
+            .wrap_err("failed to get timestamp")?;
+        let subject_client_status = state
```

Comment:
similar nits as in proto on naming here just too many things starting with `sub` makes it a slower read?